keps/sig-api-machinery/20190726-component-config-fields-should-be-optional.md

```diff
@@ -0,0 +1,197 @@
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+`ComponentConfig` allows for specification of options for a component in a
+structured and versioned way; kubernetes apimachinery objects replace flags.
+
+However, while flags make the distinction between not-specified and
+set-to-empty-value, our api types generally do not.  By using pointer-types it
+is possible to preserve this difference, this KEP proposes that we should use
+pointer-types in our componentconfig.
```

Comment:
+1! Lack of this ability has already caused problem for the scheduler. We cannot turn on pprof by default for the scheduler. In order to set the default value of "EnableProfiling" to true, we first need to find that the flag is not provided, but we cannot, because the flag is false both when it is explicitly set to false and when it is not provided. This applies to other boolean and integer options.

Reply:
@bsalamat you can tell whether a flag was provided or not using pflag flagset's `Changed` method:
https://github.com/kubernetes-sigs/addon-operators/pull/25/commits/d5f8d8d1246b33eead39aee24be1715cafb7cf5d#diff-fe99745b837e4e6db3d1c39b19f39ed3R41
That can be used for the scheduler's flag logic :+1:.

I think what is being expressed here is that is that when using a config file, once you have the struct, you won't be able to get this information unless we make all fields pointers or encode the information somewhere else.